### PR TITLE
API for injecting VMI access credentials (ssh pub keys and more)

### PR DESCRIPTION
**Pull request by davidvossel**
# Overview

The ability to dynamically inject access credentials, such as ssh public keys, into a vm guest is a common feature supported by public cloud Iaas providers. AWS, GCP, Azure, and even OpenStack all have support for propagating access credentials into VMs. Administrators can use this functionality to control access to infrastructure for both users and automation tools like ansible. It's a pretty handy feature.

In KubeVirt we don't support propagating access credentials to VMs directly. As a result, this leaves people needing to create custom cloud-init userdata scripts to inject ssh keys along with whatever host setup needs to occur. The issue with this is it couples access control of the vm with the setup of the vm. So, if someone wants to add/delete/rotate a public key, they'd have to modify every cloud-init script related to every vm that needs to pick up that credential.

That's where this PR's proposed `AccessCredentials` API comes into play. This API allows administrators to place access credentials (like ssh public keys) into a kubernetes secret, and assign that secret to a KubeVirt VMI. If the secret changes, it is propagated to the VMI's pod, which gives us a path to propagating access control changes to running VMIs.

## AccessCredentials API field

The AccessCredentials API is a list of credentials that map to kubernetes secrets. Each entry has a `type` field, which tells us what kind of credential it is, and a `propagationMethod` field which tells us how to inject the credential into the VMI. 

The goal here is to structure the API in a way that can grow with us as new functionality is added. The initial functionality provided by this PR only adds one propagation method that can be used for public ssh keys. In the future this API could be used to support things that go beyond ssh access, like windows remote access using x509 certs. 

For a working example, here's what it looks like to inject multiple ssh keys from a secret into a vm using the cloud-init configDrive provider as the propagation method. 

```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: vmi-fedora
spec:
.
.
.
  accessCredentials:
  - sshPublicKey:
      source:
        secret:
          name: my-pub-key 
      propagationMethod:
        configDrive: {}
.
.
.
```

## Dynamic Credential Propagation 

The initial propagation method introduced by this PR uses the cloud-init configDrive provider to inject ssh public keys into the VMI at boot via cloud-init metadata. This method does not support dynamic propagation into the VMI while it is running. A restart has to occur before credential changes are picked up.

By creating a API that separates where the credential comes from (k8s secret) from how it is propagated to the VMI, we open the door to much more advanced functionality in the future which can dynamically update the credential within a running guest. This could be done via an agent of some sort running in the vmi guest, a shared virtiofs directory, or some other method yet to be identified. The point is, this leaves the door open to do some pretty interesting things in the future as well as provide an initial practical working solution today with the configDrive implementation.

# Working Example

add public ssh key to secret (it's possible to add more than one key to the same secret)

```
kubectl create secret generic my-pub-key --from-file=id_rsa.pub
```

post a VMI referencing the secret as an ssh access credential. 

```
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: vmi-fedora
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
    resources:
      requests:
        memory: 1024M
  accessCredentials:
  - sshPublicKey:
      source:
        secret:
          name: my-pub-key 
      propagationMethod:
        configDrive: {}
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
    name: containerdisk
  - cloudInitConfigDrive:
      userData: |-
        #cloud-config
        password: fedora
        chpasswd: { expire: False }
    name: cloudinitdisk
```

# Example Use Cases

**Granting SSH Provisioning access used across all VM infrastructure in namespace**
- An ops team can create a single secret containing all the authorized ssh public keys their ansible provisioner uses.
- Users can create VMs and attach the standard authorization keys used to automatically provision the VMs

**Granting team member ssh access to VMs in namespace**
- All team member's public ssh keys are added to a k8s secret
- All VMs reference this secret as their ssh public key access credential

**Key rotation**
- An ops team can create a single secret containing all the authorized ssh public keys their ansible provisioner uses.
- VMs created in namespace all reference this secret which injects ssh public keys. 
- Ops team can rotate keys by updating new public keys in secret and rebooting VMs (a dynamic provisioning propagation method will be able to do this live once we implement one)

**Dynamically Granting/Revoking user access** (requires yet to be implemented propagation method that supports updating running VMIs)
- All team member's public ssh keys are added to a k8s secret
- All VMs reference this secret as their ssh public key access credential
- Adding or revoking access to VMs would be as simple as adding/removing a key from the k8s secret. This would propagate to all VMI pods and ultimately into the guest. 

```release-note
Adds new accessCredentials API field to VMI spec for injecting ssh public keys and more into vm guests
```

### CODE REVIEW

**Comment by davidvossel** [comment]
/hold

this is a WIP

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -6363,6 +6363,28 @@
      }
     }
    },
+   "v1.AccessCredential": {
+    "description": "AccessCredential represents an public authentication that can be used to authorize remote access to the vm guest",
```
Comment (chenyosef) [inline api/openapi-spec/swagger.json:6367]: an --> a

Comment (davidvossel) [reply api/openapi-spec/swagger.json:6367]: ah, thanks for pointing that out. This actually doesn't read the way I intended. I need to revise this entire sentence.

**Comment by rmohr** [comment]
> apiVersion: kubevirt.io/v1alpha3
> kind: VirtualMachineInstance
> metadata:
>   name: vmi-fedora
> spec:
> .
> .
> .
>   accessCredentials:
>     - secretName: vm-pub-keys
>       propagationMethod: configDrive
>       type: ssh
> .
> .
> .

Would it make sense to express this with structs instead of a type field? Different methods may implement different options.
How about something like this:


```
accessCredentials:
  - secretName: vm-pub-keys
    ssh:
      configDrive:
        key: val
```

another example:

```
accessCredentials:
  - secretName: vm-pub-keys
    x509:
      ignition:
        key: val
```

**Comment by davidvossel** [comment]
> Would it make sense to express this with structs instead of a type field? Different methods may implement different options.
How about something like this:

@rmohr yes. I've revised the API in a few new ways.

Now each access credential has a category "ssh, x509, etc..." and each category has both a source and propagation method. All of this is represented using structs without usage of any sort of `type` fields.

Here's a example of ssh public key pulling from a k8s secret as the source and using the configDrive propagation method. This works in the PR today.

```
  accessCredentials:
  - sshPublicKey:
      source:
        secret:
          name: my-pub-key 
      propagationMethod:
        configDrive: {}
```

Building off your previous examples, here's a "theoretical" example of a x509 cert propagated via ignition.

```
  accessCredentials:
  - x509:
      source:
        secret:
          name: my-pub-key 
      propagationMethod:
        ignition: {}
```

The main difference from your previous examples is I'm opening up the possibility of pulling credentials from sources other than secrets in the future.

Thoughts?

**Comment by rmohr** [comment]
> The main difference from your previous examples is I'm opening up the possibility of pulling credentials from sources other than secrets in the future.
> 
> Thoughts?

Works for me. I like it more than my example. My main agenda is just to have the possibility to clearly express supported combinations of propagation type and content. This is given this way.

**Comment by davidvossel** [comment]
/hold cancel

@rmohr I fleshed this out with webhook validation + unit/functional tests. The description now has some use cases as well.

File: pkg/cloud-init/cloud-init.go [hunk 2]
```diff
@@ -139,22 +163,64 @@ func ResolveNoCloudSecrets(vmi *v1.VirtualMachineInstance, secretSourceDir strin
...
+		}
+
+		for _, file := range files {
+			if file.IsDir() || strings.HasPrefix(file.Name(), "..") {
+				continue
+			}
+			keyData, keyDataError := readFileFromDir(baseDir, file.Name())
+
+			if keyDataError != nil {
+				return keys, fmt.Errorf("no public keys found at at volume: %s/%s", baseDir, file.Name())
```
Comment (ArthurSens) [inline pkg/cloud-init/cloud-init.go:203]:
```suggestion
				return keys, fmt.Errorf("no public keys found at volume: %s/%s", baseDir, file.Name())
```

**Comment by ArthurSens** [comment]
That's a really interesting PR :open_mouth: 

And I have some questions... if you don't mind answering 

I was looking at the example you gave and comparing with the [k8s documentation](https://kubernetes.io/docs/concepts/configuration/secret/#projection-of-secret-keys-to-specific-paths) on how to mount secrets within a pod.

One thing I noticed is one small difference between how you are naming the secret's name attribute

Where you have:
```
accessCredentials:
   - sshPublicKey:
       source:
         secret:
           name: my-pub-key     # name
       propagationMethod:
         configDrive: {}
```

And the example at the docs
```
  volumes:
  - name: foo
    secret:
      secretName: mysecret   # secretName
      items:
      - key: username
        path: my-group/my-username
```
If KubeVirt wants to be similar to how Kubernetes does its thing, does it makes sense to look for similar attributes names that do similar things? Maybe it would be possible to even just treat ssh pub keys as a simple volume that is being mounted by a secret

Please don't see this comment as a change request, I'm really just curious about your thought process

**Comment by davidvossel** [comment]
> If KubeVirt wants to be similar to how Kubernetes does its thing, does it makes sense to look for similar attribute's name that do similar things?

Yeah, i think changing `name` to `secretName` is a good idea. i'll make this change.

> Maybe it would be possible to even just treat ssh pub keys as a simple volume that is being mounted by a secret

at least on the pod level, that's what's happening. behind the scenes we're mapping the secret as a volume to the Pod. virt-launcher is then taking that secret passed as a volume to the pod and injecting it into the vm guest.

**Comment by xpivarc** [comment]
@davidvossel Do we have docs for this?

File: pkg/cloud-init/cloud-init.go [hunk 3]
```diff
@@ -263,18 +329,19 @@ func readCloudInitConfigDriveSource(source *v1.CloudInitConfigDriveSource) (*Clo
...
 		LocalHostname: hostname,
 	}
 }
 
-func readCloudInitConfigDriveMetaData(uid, name, hostname, namespace string) *Metadata {
-	return &Metadata{
-		UUID:       uid,
-		InstanceID: fmt.Sprintf("%s.%s", name, namespace),
-		Hostname:   hostname,
+func readCloudInitConfigDriveMetaData(uid, name, hostname, namespace string, keys map[string]string) *ConfigDriveMetadata {
```
Comment (xpivarc) [inline pkg/cloud-init/cloud-init.go:339]: I know this is not your change but why this is called `readCloudInitConfigDriveMetaData`?

File: pkg/cloud-init/cloud-init.go (hunk 2 above)
Comment (davidvossel) [reply pkg/cloud-init/cloud-init.go]: fixed

File: pkg/cloud-init/cloud-init.go [hunk 4]
```diff
@@ -139,22 +163,64 @@ func ResolveNoCloudSecrets(vmi *v1.VirtualMachineInstance, secretSourceDir strin
...
+			return keys, err
+		}
+
+		for _, file := range files {
+			if file.IsDir() || strings.HasPrefix(file.Name(), "..") {
+				continue
+			}
+			keyData, keyDataError := readFileFromDir(baseDir, file.Name())
+
+			if keyDataError != nil {
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/cloud-init/cloud-init.go]: fixed, this was probably a result of pulling code from one function into another.

File: pkg/cloud-init/cloud-init.go [hunk 5]
```diff
@@ -139,22 +163,64 @@ func ResolveNoCloudSecrets(vmi *v1.VirtualMachineInstance, secretSourceDir strin
...
+	count := 0
+	for _, accessCred := range vmi.Spec.AccessCredentials {
+
+		// check to see if access credential is propagated by config drive or not
+		if accessCred.SSHPublicKey == nil || accessCred.SSHPublicKey.PropagationMethod.ConfigDrive == nil {
+			continue
+		}
+
+		secretName := ""
+		if accessCred.SSHPublicKey != nil && accessCred.SSHPublicKey.Source.Secret != nil {
```
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/cloud-init/cloud-init.go]: yep, fixed

**Comment by davidvossel** [comment]
/hold

I'd like to further investigate a supportable option for dynamic key propagation

**Comment by fabiand** [comment]
Should this API also support injecting local/console access credentials?

Speak passwd/shadow changes.

**Comment by davidvossel** [comment]
i'm reworking this.

**Pull request closed without merging**